packages/nodes-base/nodes/Mcp/McpServer.ts

```diff
@@ -0,0 +1,199 @@
+ * call. This is necessary in order to not have executions for listing tools
+ * and other commands sent by the MCP client
+ */
+function wasToolCall(body: string) {
+	try {
+		const message: unknown = JSON.parse(body);
+		const parsedMessage: JSONRPCMessage = JSONRPCMessageSchema.parse(message);
+		return (
+			'method' in parsedMessage &&
+			'id' in parsedMessage &&
+			parsedMessage?.method === CallToolRequestSchema.shape.method.value
+		);
+	} catch {
+		return false;
+	}
+}
+
+export class McpServer {
+	servers: { [sessionId: string]: Server } = {};
+
+	transports: { [sessionId: string]: FlushingSSEServerTransport } = {};
+
+	logger: Logger;
+
+	private tools: { [sessionId: string]: Tool[] } = {};
+
+	private resolveFunctions: { [sessionId: string]: CallableFunction } = {};
+
+	constructor(logger: Logger) {
+		this.logger = logger;
+		this.logger.debug('MCP Server created');
+	}
+
+	async connectTransport(postUrl: string, resp: CompressionResponse): Promise<void> {
+		const transport = new FlushingSSEServerTransport(postUrl, resp);
+		const server = this.setUpServer();
+		const { sessionId } = transport;
+		this.transports[sessionId] = transport;
+		this.servers[sessionId] = server;
+
+		resp.on('close', async () => {
+			this.logger.debug(`Deleting transport for ${sessionId}`);
+			delete this.tools[sessionId];
+			delete this.resolveFunctions[sessionId];
+			delete this.transports[sessionId];
+			delete this.servers[sessionId];
+		});
+
+		await server.connect(transport);
+
+		// Make sure we flush the compression middleware, so that it's not waiting for more content to be added to the buffer
+		if (resp.flush) {
+			resp.flush();
+		}
+	}
+
+	async handlePostMessage(req: express.Request, resp: CompressionResponse, connectedTools: Tool[]) {
+		const sessionId = req.query.sessionId as string;
+		const transport = this.transports[sessionId];
+		this.tools[sessionId] = connectedTools;
+		if (transport) {
+			// We need to add a promise here because the `handlePostMessage` will send something to the
+			// MCP Server, that will run in a different context. This means that the return will happen
+			// almost immediately, and will lead to marking the sub-node as "running" in the final execution
+			await new Promise(async (resolve) => {
+				this.resolveFunctions[sessionId] = resolve;
+				await transport.handlePostMessage(req, resp, req.rawBody.toString());
+			});
+			delete this.resolveFunctions[sessionId];
+		} else {
+			this.logger.warn(`No transport found for session ${sessionId}`);
+			resp.status(401).send('No transport found for sessionId');
+		}
+
+		if (resp.flush) {
+			resp.flush();
+		}
+
+		delete this.tools[sessionId]; // Clean up to avoid keeping all tools in memory
+
+		return wasToolCall(req.rawBody.toString());
+	}
+
+	setUpServer(): Server {
+		const server = new Server(
+			{
+				name: 'n8n-mcp-server',
+				version: '0.1.0',
+			},
+			{
+				capabilities: { tools: {} },
+			},
```

Comment:
Do we need this? `ServerCapabilities` seems to be optional

Reply:
looks like we do. removing this makes the SDK throw errors saying `Server does not support tools`